cookbook/integrations/flytekit_plugins/greatexpectations/README.rst

```diff
@@ -0,0 +1,68 @@
+Great Expectations
+==================
+
+**Great Expectations** is a Python-based open-source library for validating, documenting, and profiling your data. 
+It helps to maintain data quality and improve communication about data between teams.
+
+The goodness of data validation in Great Expectations can be integrated with Flyte to validate the data moving in and out of 
+the pipeline entities you may have defined in Flyte. This helps establish stricter boundaries around your data to 
+ensure that everything's as you expected and that data isn't going to crash your pipelines anymore unexpectedly!
+
+How to Define Your Integration
+------------------------------
+
+Great Expectations accepts datasets that can only be "strings" -- file names, table names, etc. 
```

Comment:
Oh, I'm sorry! I was referring to `data_asset`, not the Datasource. Rephrased it now. Let me know if the update looks okay. :)

Reply:
No worries! This has also changed a bit over time. I am still not sure whether this is entirely accurate, but I also may be misunderstanding some things here. 

To clarify: 
If your Datasource is a directory of csvs, or a SQL table, then the statement here is correct - you would specify the csv file or table as a string with an `Inferred` or `ConfiguredAssetDataConnector`, and then the DataConnector would help retrieve the object with that name. 

However, you could also use a RuntimeDataConnector to pass in a pandas or Spark DataFrame, in which case you would be passing the DataFrame object directly into the BatchRequest. In this case, Great Expectations still requires you to specify a `data_asset_name`, but this is essentially an arbitrary string, for users to help keep track of the data used. It is this latter case of RuntimeDataConnectors which I think might not be accurately depicted here.

Am I understanding this correctly? It is totally possible that I might be missing something here!

Reply:
@talagluck, I think I got your point. We have the privilege to give a non-string (DataFrame) when using a `RuntimeDataConnector` but not when using an `Inferred` or a `ConfiguredAssetDataConnector` (correct?).

I have a question here. When I was surfing through the examples in Great Expectations' Docs, I've come across this example:

```python
df: pd.DataFrame = pd.read_csv("some_path.csv")
runtime_batch_request = RuntimeBatchRequest(
    datasource_name="my_pandas_datasource",
    data_connector_name="my_runtime_data_connector",
    data_asset_name="insert_your_data_asset_name_here",
    runtime_parameters={
      "batch_data": df
    },
    batch_identifiers={
        "some_key_maybe_pipeline_stage": "ingestion step 1",
        "some_other_key_maybe_airflow_run_id": "run 18"
    }
)
```
(I believe this is what you're referring to)

`runtime_parameters` has `batch_data` which allows us to give a DataFrame object. However, there's this line in the changelog saying that `batch_data` isn't available anymore: https://docs.greatexpectations.io/en/latest/changelog.html?highlight=RuntimeDataConnector#id9

So now, what's the correct way of passing in a DataFrame object? Can you point me to an example if there's any? Also, this is the type I've attached to `runtime_parameters` in the plugin:

```python
runtime_parameters: typing.Optional[typing.Dict[str, str]] = None
```
Should this be modified to `runtime_parameters: typing.Optional[typing.Dict[str, typing.Any]] = None`?

Reply:
Hi @samhita-alla

> We have the privilege to give a non-string (DataFrame) when using a RuntimeDataConnector but not when using an Inferred or a ConfiguredAssetDataConnector (correct?).

This is exactly right.

> runtime_parameters has batch_data which allows us to give a DataFrame object. However, there's this line in the changelog saying that batch_data isn't available anymore: https://docs.greatexpectations.io/en/latest/changelog.html?highlight=RuntimeDataConnector#id9

I understand the confusion here. `batch_data` was removed from the `BatchRequest` object because we added a `RuntimeBatchRequest` object that uses it instead. It appears that your code in the Flyte integration already takes into account the possibility of using a `RuntimeBatchRequest` through the `BatchRequestConfig` object, so I don't know if much will need to be updated there. [You can see an example of using a `RuntimeBatchRequest` in this way here](https://docs.greatexpectations.io/en/latest/guides/how_to_guides/creating_batches/how_to_load_a_pandas_dataframe_as_a_batch.html) under the Show Docs for V3 tab.

> Also, this is the type I've attached to runtime_parameters in the plugin:

> runtime_parameters: typing.Optional[typing.Dict[str, str]] = None

> Should this be modified to runtime_parameters: typing.Optional[typing.Dict[str, typing.Any]] = None?

Yes - this looks right in light of the ability to pass in a DataFrame here.

Thanks for the follow-up here - some great questions!

Reply:
@talagluck: Can we not have an automated mechanism in Great Expectations wherein `data_asset_name` is automatically picked by `runtime_parameters` if the batch request is a `RuntimeBatchRequest`? This is because `data_asset_name` doesn't add any value if we're using a `RuntimeBatchRequest`.

(Say) In the case of DataFrames, you can populate the `batch_data`, and in the case of SQL queries, you can populate the table name.

What do you think?

Reply:
Okay, thank you!